tfprotov5/tf5server/server.go

```diff
@@ -130,61 +184,170 @@ func (s *server) stoppableContext(ctx context.Context) context.Context {
 	return stoppable
 }
 
+// loggingContext returns a context that wraps `ctx` and has
+// terraform-plugin-log loggers injected.
+func (s *server) loggingContext(ctx context.Context) context.Context {
+	// generate a request ID
+	reqID, err := uuid.GenerateUUID()
+	if err != nil {
+		reqID = "unable to assign request ID: " + err.Error()
+	}
+
+	// set up the logger SDK loggers are derived from
+	ctx = tfsdklog.NewRootSDKLogger(ctx, append(tfsdklog.Options{
+		tfsdklog.WithLevelFromEnv("TF", "LOG", "SDK"),
+	}, s.tflogSDKOpts...)...)
+	ctx = tfsdklog.With(ctx, "tf_req_id", reqID)
+	ctx = tfsdklog.With(ctx, "tf_provider_addr", s.name)
```

Comment:
Open question: since these could be considered part of the compatibility for downstream handling, such as practitioner/developer filtering and tooling, should these be constants and/or documented anywhere?

Reply:
I think my hot take is we don't offer any compatibility guarantees at the moment, so I don't know that there's anything to document (except maybe that stance, which is [already documented in the `TF_LOG=JSON` documentation](https://www.terraform.io/docs/internals/debugging.html)) and I don't know that a constant makes sense to me. The benefits of constants, in my mind, are:

* People can import them to use that value in their own code. I don't see a use case for this under the compatibility guarantees we offer at the moment, and would actually want to dissuade people from doing this at the moment, at least.
* We can ensure consistency in that value in the code. But because these values only are being used in this specific block and are right next to each other, I think the cost of indirection ("wait, what was that value getting logged as? let me scroll up to the top of the file and find it") isn't outweighed by the benefit of enforcing consistency here.

Is there another benefit we're after by making these constants that I'm missing?

Reply:
Constants can have code comments (`// this will contain data such as "terraform-provider-example"`) so we know if we should reuse this logging key and constants can be grouped together (do they all follow a similar naming patterns or are some `_id` and others `_identifier`?). They don't need to be exported.

Reply:
I think these lines can also have code comments, too? And these lines are grouped together by use? I guess I'm still not following on what we're trying to achieve by making them constants.

Reply:
There are logging keys spread out in a few places already, I'm seeing:

```
"tf_req_id"
"tf_provider_addr"
"tf_rpc"
"tf_resource_type"
"tf_data_source_type"
"tf_proto_version"
```

It's not clear to me why some names are shortened and what they contain without further reading or knowing the code. Keeping them in a constants section (or its own file), can make it clear to someone who needs to better understand the expected value and what's available (or not). Then writing the website document is just looking in one place. We can even say, "if you touch this, update the website", e.g.


```go
// Logging keys used in this code.
// Any updates should be reflected on the documentation website.
// Please note that practitioners or tooling reading logs may be depending on these keys, so be cautious about changing them.
const (
  // Terraform request identifier. Its value is a UUIDv4 generated by this code during protocol requests.
  logKeyTfReqId = "tf_req_id"

  // Terraform provider address. Its value is the name of the provider, such as "terraform-provider-example".
  logKeyTfProviderAddr = "tf_provider_addr"

  // Terraform remote procedure call. Its value is the name of the call, such as "ApplyResourceChange".
  logKeyTfRpc = "tf_rpc"

  // Terraform resource type, such as, "random_pet".
  logKeyTfResourceType = "tf_resource_type"

  // Terraform data source type, such as, "archive_file".
  logKeyTfDataSourceType = "tf_data_source_type"

  // Terraform protocol version, such as, "5".
  logKeyTfProtoVersion = "tf_proto_version"
)
```

Reply:
I still don't know that I agree. I find the editor argument compelling, but these are also things that shouldn't be appearing in the code more than a handful of times and really shouldn't be typed often. Further, the SDKs are going to have a bunch of these one-off keys that are function-specific and only used once or _maybe_ twice, and that's a lot of indirection that makes reading the code harder. It also makes grepping the code into a two step process; find the variable, then find uses of the variable.

I think moving these to constants makes it _harder_ to see what they contain; when they're in constants, you need to hope someone wrote a comment, that the comment is detailed enough, and that the comment is up to date. Are we sure that's going to be the case? When they're inline, what they contain is literally right next to them.

I think documenting every log field available on the website is a non-goal, and the ones we do document should be documented to support a specific use case, because otherwise we're creating a whole compatibility interface that's going to constrain us in the future for dubious benefit. I think we may have different expectations of how people should be interacting with these logs and what kind of interface they represent, and I think that's a good conversation we should have, but we should probably find a better time and venue for it.

This just feels kinda like using constants for map keys to me. Yes, it's a string, yes, it's used two or even three times, but that doesn't automatically make it read as a constant to me, and I think there are costs with making it a constant. I find the argument about editors compelling, I find the argument about enforcing consistency compelling, but I don't know that either of those benefits outweighs the costs associated with constants, like increased indirection, and the difficulty of adding new log lines or new variables to existing log lines if we're going to assert this as our general practice. I'm still not following why the comments being over constants instead of over log lines is better or clearer. I'm very wary of the possibility of having hundreds of constants that you need to refer to and grep through and add to when adding a log line, because I think that reduces the likelihood we'll surface information, not because it isn't useful, but because surfacing it is no longer trivial. I'm wondering if we're envisioning the use of this logging similarly, where this PR is just the start and over time will come to surface more information from the SDKs, like variable values and such when making decisions, or if that's where our disconnect is.

All that said, I think the impact of the immediate change being discussed isn't likely to exceed the impact of holding the PR for it, so I'm gonna go ahead and run with this, but I don't know that I'm sold on this being a general practice or The Way We Do Things. I think I'd like to continue that conversation later.